[nus-jeff] iP

nus-jeff opened the pull request:
# D U K E #
> "*Duke is the best ~~partner~~ task tracker I have ever had!!*" - nobody ([hyperlink](https://www.youtube.com/watch?v=dQw4w9WgXcQ&ab&ab))
---
Start using Duke today, it is very **worth it** as it is:
+ Easy to use!
+ Super easy to use!!
+ Extremely easy to use!!!
---
Duke comes with the **ability** to:
1. Add or Delete tasks
2. Mark tasks as done or not done
3. Search for all tasks with a keyword
---
More features to be implemented:
- [x] Ability to exit Duke
- [ ] Ability to talk to Duke :kissing_heart: (coming soon!)
---
Here is a preview of our `Ui` in java code:
```Modelica
public static void welcomeMessage() {
    String logo = " ____        _        \n"
            + "|  _ \\ _   _| | _____ \n"
            + "| | | | | | | |/ / _ \\\n"
            + "| |_| | |_| |   <  __/\n"
            + "|____/ \\__,_|_|\\_\\___|\n";
    System.out.println("Hello from\n" + logo);
    System.out.println(DIVIDER_START_END
            + "What can the Duke help you with today?\n"
            + DIVIDER_START_END);
}
```

Commit:
Add Gradle support

nus-jeff committed:
Added greeting, echoing of input and goodbye message.

nus-jeff committed:
added todo list function, modified gitignore to remove extra text

nus-jeff committed:
added Task class, implemented mark/unmark function

nus-jeff committed:
added ToDoTask, DeadlineTask, EventTask

nus-jeff committed:
Edit test cases, to test

nus-jeff committed:
Refactor code to use ArrayList, update test case expected output

nus-jeff committed:
Add delete function to Duke, update test case for it

nus-jeff committed:
Add DukeException, refactor code for error handling, numbersFormatException to be handled in future

nus-jeff committed:
Add file save ability to ToDoList, file path to fix

nus-jeff committed:
fix file path issue with ToDoList

nus-jeff committed:
Add date formatting for deadline & event case

nus-jeff committed:
Change location of save file to project file as root rather than user home

nus-jeff committed:
Remove println statement from debugging

nus-jeff committed:
Merge branch 'branch-level-7'

nus-jeff committed:
Resolve merge conflict from branch-level-8

nus-jeff committed:
Add missing abstract save method in Task, add getToDoCount in ToDoList
~

nus-jeff committed:
Fix errors in saving and loading of ToDoList, test case update with new input

nus-jeff committed:
Add Storage class to deal with loading and saving or ToDoList

JamesLiuZX commented on src/main/java/ToDoList.java:
```diff
@@ -0,0 +1,133 @@
+                + " - " + arr.get(ind) + "\n" + divider);
+    }
+
+    public void save() throws IOException {
```
Generally, for error handling, there are 2 types: adding a `throws` clause to a method, or using a try-catch block. The first one should only be used if you are throwing it upwards to be handled by the consumer method,  or caused by input parameters, or you don't want to handle them and just let the program stop. In all other cases, the try-catch block is preferred as you can meaningfully handle the exceptions and let users understand the error.

JamesLiuZX commented on src/main/java/EventTask.java:
```diff
@@ -0,0 +1,39 @@
+        return "EVENT/+/" + status + this.title + "/+/" + this.from + "/+/" + this.to + "\n";
+    }
+
+    @Override
+    public String toString() {
+        String status = this.isDone ? "[x] " : "[ ] ";
+        return "[E]" + status + this.title + " (from: " + this.from + " to: " + this.to + ")";
```
Would be cleaner if you used String.format()

JamesLiuZX commented on src/main/java/ToDoList.java:
```diff
@@ -0,0 +1,133 @@
+
+
+public class ToDoList {
+    private ArrayList<Task> arr = new ArrayList<>(); //1-based index
```
Generally, even for user inputs that start with 1-index. You should use 0-index for your internal data structures and convert the user inputs index. Otherwise you'll have to do a lot of necessary steps down the road with your methods such as deleting, checking for max length, and your constructor having to add a dummy task which takes up resources.

JamesLiuZX commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+                    + "| |_| | |_| |   <  __/\n"
+                    + "|____/ \\__,_|_|\\_\\___|\n";
+            System.out.println("Hello from\n" + logo);
+            String divider = "____________________________________________________________\n";
+            System.out.println(divider + "What can the Duke help you with today?\n" + divider);
```
If there's more than one instance, you can initialise the divider String as a private variable within your class.

JamesLiuZX commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+        }
+    }
+
+    private static String[] commandHandler(String input, String regex,int limit, int minSize) throws DukeException {
```
Note proper spacing between arguments.

JamesLiuZX commented on src/main/java/DeadlineTask.java:
```diff
@@ -0,0 +1,36 @@
+
+    private String formatIfDate(String input, LocalDate date) {
+        DateTimeFormatter inputFormat = DateTimeFormatter.ofPattern("d/M/uuuu");
+        DateTimeFormatter outputFormat = DateTimeFormatter.ofPattern("dd LLL uuuu");
```
Would be clearer if you explained the use of 'L' and 'u' instead of M and Y

JamesLiuZX commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,108 @@
+                String command = input[0];
+                String sub;
+
+                switch (command) {
+                    case "bye":
+                        return;
+                    case "list":
+                        System.out.println(ls);
+                        break;
+                    case "mark":
+                        sub = input[1];
+                        ls.markTask(Integer.parseInt(sub)); //numbersformatexception to be handled
+                        break;
+                    case "unmark":
+                        sub = input[1];
+                        ls.unmarkTask(Integer.parseInt(sub)); //numbersformatexception to be handled
+                        break;
+                    case "delete":
+                        sub = input[1];
+                        ls.delete(Integer.parseInt(sub)); //numbersformatexception to be handled
+                        break;
+                    case "todo":
+                    case "event":
+                    case "deadline":
+                        Duke.taskCommandHandler(input, ls);
+                        break;
+                    default:
+                        throw new DukeException("The Duke does not understand your words!");
```
Good job making use of exceptions in default!

JamesLiuZX replied:
Could create a new specific exception for this.

JamesLiuZX replied:
As per Java coding standards, the cases should be on the same indentation level as switch.

JamesLiuZX reviewed (commented):
Overall great work! The code logic is great, and I can see good use of OOP. Keep it up!

nus-jeff committed:
Add UI class to deal with displays

nus-jeff committed:
Add Parser class to deal with inputs

nus-jeff committed:
Create packages and organize the different classes into packages

swxk19 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Scanner;
+
+public class Duke {
+    public static void main(String[] args) {
```
I like how you abstracted out all the `switch-case` handling of the commands out of `main`. Makes it look short and sweet!

swxk19 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,65 @@
+        if (ind < 1 || ind > toDoCount) {
+            throw new IndexDukeException();
+        }
+        Task rm = arr.get(ind);
```
Variable name can be more descriptive, perhaps `taskToRemove`

swxk19 commented on src/main/java/duke/functions/Parser.java:
```diff
@@ -0,0 +1,86 @@
+
+public class Parser {
+    private  Parser() {
+    }
```
Constructor may not need to be defined if you are just using the default constructor :)

swxk19 reviewed (commented):
Nice Job!

nus-jeff committed:
Merge remote-tracking branch 'origin/add-gradle-support'

nus-jeff committed:
Add JUnit test case for Parser class and ToDoList class

nus-jeff committed:
Fix wrong UI output due to package refactor, update EXPECTED.TXT to correctly match

nus-jeff committed:
Add find function to ToDoList, test case now includes find function

nus-jeff committed:
Refactor getTask method in ToDoList to prevent access to index 0 Task, getTask now throws IndexDukeException

nus-jeff committed:
Refactor Storage class to allow storing of any type of path, path stored in storage are now with reference to user home directory

nus-jeff committed:
Modify listMsg method in UI class to take in a ToDoList instead of its String representation

nus-jeff committed:
Add javadoc header comments to public classes and methods in src/main/java/duke

nus-jeff committed:
Refactor code to meet conding standards

nus-jeff committed:
Refactor code to meet conding standards

nus-jeff committed:
Merge branch 'branch-level-9'

nus-jeff committed:
Resolve merge conflict, tidy up code to meet coding standard

nus-jeff committed:
Fix merge conflict

nus-jeff committed:
Add template JavaFX based on tutorial

nus-jeff committed:
Add GUI for Duke, refactor code for GUI, update test cases

nus-jeff committed:
Merge branch 'branch-level-10'

nus-jeff committed:
Fix error in test cases

nus-jeff committed:
Add java assertions to code, add IntParseDukeException class

Some parts of the code relies on the programmer entering
the correct input argument in order for the code to not
exhibit any unwanted behaviour.

Let's use java assertion to check for such unwanted
behaviours caused by the programmer's wrong input.

Exceptions may also be thrown while trying to parse
a string input that is not an integer for the mark,
unmark and delete command in Duke.

Let's handle this exception by creating an
IntParseDukeException that gets throwns when such
an error occurs.

nus-jeff committed:
Add reminder to Duke and LoadDukeException class.

Users have to manually look at their to do list in
Duke in order to find any tasks that is about to happen.

Let's create a reminder that appears everytime Duke starts
up to remind users that the given task is about to happen.

The number of days to the task to be compared to is set at
10 days. The ability to change this day will be implemented
in the future.

A new DukeException class, LoadDukeException has also been
created to deal with errors arising from loading the file
from the given path to Duke.

Code has been refactored to make the reminder function work
and to throw and catch LoadDukeException.

nus-jeff committed:
Improve code quality for EventTask and DeadlineTask class

nus-jeff committed:
Improve code quality

Certain parts of the code are of low quality
such as the existance of arrow-head style
code and empty catch blocks.

Let's improve the quality of code by refactoring
it. Changes include but are not limited to:
+ extracting parts of code to another method
+ comments to explain empty catch blocks

nus-jeff committed:
Merge branch 'master' into branch-A-Assertions

nus-jeff committed:
Merge pull request #1 from nus-jeff/branch-A-Assertions

Add java assertions to code, add IntParseDukeException class

nus-jeff committed:
Fix merge conflict

nus-jeff committed:
Merge pull request #2 from nus-jeff/branch-A-CodeQuality

Improve Code Quality

nus-jeff committed:
Touch up code based on SE Bot suggestion

nus-jeff committed:
Edit the Gui to look slightly nicer

nus-jeff committed:
Update README.md

nus-jeff committed:
Update README.md

nus-jeff committed:
Update README.md

nus-jeff committed:
Update README.md

nus-jeff committed:
Update README.md

nus-jeff committed:
Update README.md

nus-jeff committed:
Add Ui.png

nus-jeff committed:
Add title to program window, update Ui.png to show it

nus-jeff committed:
Remove Ui.png

nus-jeff committed:
Add  Ui.png

nus-jeff committed:
Update README.md
